File: 2581. Count Number of Possible Root Nodes/Solution.java
class Solution {
    List[] graph;
    Map<Integer, Set<Integer>> g;
    int[] parent;
    int res;
    public int rootCount(int[][] edges, int[][] guesses, int k) {
        int n = edges.length + 1, count = 0;
        res = 0;
        parent = new int[n];
        graph = new List[n];
        g = new HashMap<>();
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
            g.put(i, new HashSet<>());
        }
        g.put(-1, new HashSet<>());
        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }
        for (int[] guess : guesses) g.get(guess[0]).add(guess[1]);
        fillParent(0, -1);
        for (int i = 1; i < n; i++) {
            if (g.get(parent[i]).contains(i)) count++;
        }
        if (count >= k) res++;
        List<Integer> neighbours = graph[0];
        for (int i : neighbours) dfs(i, 0, count, k);
        return res;
    }
    public void fillParent(int curr, int prev) {
        List<Integer> neighbours = graph[curr];
        parent[curr] = prev;
        for (int neighbour : neighbours) {
            if (neighbour == prev) continue;
            fillParent(neighbour, curr);
        }
    }
    public void dfs(int curr, int prev, int count, int k) {
        if (g.get(curr).contains(prev)) count++;
        if (g.get(prev).contains(curr)) count--;
        if (count >= k) res++;
        List<Integer> neighbours = graph[curr];
        for (int neighbour : neighbours) {
            if (neighbour == prev) continue;
            dfs(neighbour, curr, count, k);
        }
    }
}

File: 2581. Count Number of Possible Root Nodes/Solution.py
class Solution:
    def rootCount(self, edges: List[List[int]], guesses: List[List[int]], k: int) -> int:
        n, count, self.res = len(edges) + 1, 0, 0
        self.graph = [[] for i in range(n)]
        self.guess = [set() for i in range(-1, n)]
        self.parent = [-1 for i in range(n)]
        for x, y in edges:
            self.graph[x].append(y)
            self.graph[y].append(x)
        for x, y in guesses:
            self.guess[x].add(y)
        self.fillParent(0, -1)
        for i in range(n):
            if i in self.guess[self.parent[i]]: count += 1
        if count >= k: self.res += 1
        for neighbour in self.graph[0]: self.dfs(neighbour, 0, count, k)
        return self.res
    
    def fillParent(self, curr, prev):
        self.parent[curr] = prev
        for neighbour in self.graph[curr]:
            if neighbour == prev: continue
            self.fillParent(neighbour, curr)
    
    def dfs(self, curr, prev, count, k):
        if prev in self.guess[curr]: count += 1
        if curr in self.guess[prev]: count -= 1
        if count >= k: self.res += 1
        for neighbour in self.graph[curr]:
            if neighbour == prev: continue
            self.dfs(neighbour, curr, count, k)